### PR TITLE
Jobs on GPU

### PR DESCRIPTION
**Pull request by hhsecond**
- CircleCI config for running GPU jobs
- Test cases running on GPU Device

I am not using any of the existing build utilities such as system-setup.py since relying on a file inside the repo trigger rebuilding of that layer on each push which is time-consuming. We need to iterate over this and other build processes to unify the build processes and to do the packaging and release of GPU docker image etc

Also, because of the `allow growth` feature in TF, the test cases on GPU takes a lot of time now

### CODE REVIEW

File: test/basic_tests.py [hunk 1]
```diff
@@ -13,13 +11,15 @@
 import sys
 
 sys.path.insert(0, os.path.join(os.path.dirname(__file__), "../opt/readies"))
-import paella
```
Comment (rafie) [inline test/basic_tests.py]: Better to keep it. It is useful for debugging purposes (enables easily debugger invocation and insertion of breakpoints in code).

File: opt/Makefile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 
-ROOT:=..
+ROOT:=$(abspath ..)
```
Comment (rafie) [inline opt/Makefile]: No need for abspath here as readies/mk/main will abspath it.

File: Dockerfile.gpu-test [hunk 3]
```diff
@@ -0,0 +1,65 @@
+FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu18.04 AS builder
+
+ENV DEPS "build-essential git ca-certificates curl unzip wget libgomp1 patchelf coreutils tcl libjemalloc-dev"
```
Comment (rafie) [inline Dockerfile.gpu-test]: We should consider utilizing /opt/system-setup.py infrastructure for requirements installations, as we repeat them in CircleCI and native environments.

File: .circleci/config.yml [hunk 4]
```diff
@@ -90,7 +90,24 @@ jobs:
...
+      - checkout
+      - run:
+          name: Bulding Docker
+          command: |
+            docker build -f Dockerfile.gpu-test -t redisai-gpu-test .
+      - run:
+          name: Test
+          command: |
+            mkdir -p ~/workspace/tests
+            docker run --gpus all -v $HOME/workspace/tests:/data/tests/logs -it --rm redisai-gpu-test
```
Comment (rafie) [inline .circleci/config.yml]: I'm not sure whether CircleCI supports volumes (-v) with docker.

File: .circleci/config.yml [hunk 5]
```diff
@@ -90,7 +90,24 @@ jobs:
...
-
+  build-gpu:
+    machine:
+      docker_layer_caching: true
+      resource_class: gpu.nvidia.medium
+      image: ubuntu-1604-cuda-10.1:201909-23
+    steps:
+      - checkout
+      - run:
+          name: Bulding Docker
```
Comment (rafie) [inline .circleci/config.yml]: "Building"

File: test/basic_tests.py (hunk 1 above)
Comment (hhsecond) [reply test/basic_tests.py]: ok

**Comment by hhsecond** [comment]
@lantiga @rafie I agree with you on moving the dependency installations to `system-setup.py` but the fact that docker layer caching doesn't work on `ADD` for some reason causes all of the dependency installation to reruns on all the builds. This seems inefficient for me especially since GPU resources are costly. Let's see if CircleCI folks have some idea about this and I'll move to `system-setup` anyways

File: .circleci/config.yml (hunk 4 above)
Comment (hhsecond) [reply .circleci/config.yml]: Just verified. It works on `machine` executor

**Comment by hhsecond** [comment]
@lantiga @rafie 
I have removed all the dependency installations from the docker file and relying on `system-setup.py` for most of the part. But dependencies required for the Redis installations are still in the docker file. It is because docker layer caching we had enabled wasn't working as expected and putting those dependencies also to system-setup will cause the job to take quite a lot of time to finish. I have raised a complaint with them already. Let see what they have got.
Also, the GPU backend is disabled in circle CI from today and they are going to send us a contract soon.
LET'S NOT MERGE THIS TILL THE GPU SUPPORT IS BACK

**Comment by lantiga** [comment]
@hhsecond Thank you Sherin, great job. Given the costs associated with building on GPU, it's really important we don't waste precious compute minutes building.

File: opt/build/libtorch/repack.sh [hunk 6]
```diff
@@ -62,6 +65,7 @@ fi
 LIBTORCH_ZIP=libtorch-${PT_BUILD}-${PT_OS}-${PT_ARCH}-${PT_VERSION}.zip
 if [ ! -f $LIBTORCH_ARCHIVE ]; then
 	echo "Downloading libtorch ${PT_VERSION} ${PT_BUILD}"
+  echo $LIBTORCH_URL
```
Comment (rafie) [inline opt/build/libtorch/repack.sh]: echo can be removed.

File: opt/build/libtorch/repack.sh [hunk 7]
```diff
@@ -10,13 +10,12 @@ ROOT=$HERE/../../..
...
-	GPU=no
+	GPU=0
 elif [[ "$1" == "gpu" ]]; then
-	GPU=yes
+	GPU=1
 else
-	GPU=${GPU:-no}
+	GPU=${GPU:-0}
 fi
-
```
Comment (rafie) [inline opt/build/libtorch/repack.sh]: We can keep the blank line. :-)

File: test/test_requirements.txt [hunk 8]
```diff
@@ -1,3 +1,4 @@
 numpy
+scipy
```
Comment (rafie) [inline test/test_requirements.txt]: Do we really need scipy?

File: get_deps.sh [hunk 9]
```diff
@@ -169,7 +169,8 @@ if [[ $WITH_TFLITE != 0 ]]; then
 
 		LIBTFLITE_ARCHIVE=libtensorflowlite-${TFLITE_OS}-${TFLITE_ARCH}-${TFLITE_VERSION}.tar.gz
 
-		[[ ! -f $LIBTFLITE_ARCHIVE || $FORCE == 1 ]] && wget --quiet $LIBTF_URL_BASE/$LIBTFLITE_ARCHIVE
+		[[ ! -f $LIBTFLITE_ARCHIVE || $FORCE == 1 ]] && wget $LIBTF_URL_BASE/$LIBTFLITE_ARCHIVE
+		echo $LIBTF_URL_BASE/$LIBTFLITE_ARCHIVE
```
Comment (rafie) [inline get_deps.sh]: echo can be removed.

File: test/test_requirements.txt (hunk 8 above)
Comment (lantiga) [reply test/test_requirements.txt]: I don’t think we do

**Comment by lantiga** [comment]
(@hhsecond please address Rafi’s comments before merging)

**Comment by lantiga** [comment]
Hey @hhsecond, are we good to go?

**Pull request merged**